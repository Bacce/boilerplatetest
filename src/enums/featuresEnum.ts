export enum featuresEnum {
    FEATURE_1="FEATURE_1",
    FEATURE_2="FEATURE_2",
    FEATURE_3="FEATURE_3",
    FEATURE_4="FEATURE_4",
  }