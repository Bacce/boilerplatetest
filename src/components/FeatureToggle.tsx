import { useEffect, useState } from "react";
import { featuresEnum } from "../enums/featuresEnum";
import { wChild } from "../types";

interface IFeatureToggleProps extends wChild {
  feature: featuresEnum;
  offTemplate?: JSX.Element | JSX.Element[] | React.ReactChild;
}

// Do not render children if the tag is on the disabled list
export const FeatureToggle = ({children, feature, offTemplate}:IFeatureToggleProps) => {
  const [featureList, setFeatureList] = useState<string[]>([]);

  useEffect(()=>{
    try {
      const lsData = localStorage.getItem("jt-disabled-features");
      if(lsData) setFeatureList(JSON.parse(lsData));
    }
    catch(e){}
  },[]);

  return featureList.indexOf(feature) > -1 ? <>{offTemplate && offTemplate}</> : <>{children}</>;
}

// Check if feature toggle configuration is available, and populates it with default
export const FeatureToggleProvider = ({children}:wChild) => {
  useEffect(() => {
    try { // In case the localstorage is disabled
      const lsData = localStorage.getItem("jt-disabled-features");
      if(!lsData) {
        let defaultValue = process.env.REACT_APP_DISABLED_FEATURES;
        if(!defaultValue) defaultValue = "[]";
        localStorage.setItem("jt-disabled-features", defaultValue);
      }
    } catch (e) {}
  },[]);
  return (<>{children}</>);
}