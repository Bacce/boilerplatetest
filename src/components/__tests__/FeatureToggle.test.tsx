import { render } from '@testing-library/react';
import { featuresEnum } from '../../enums/featuresEnum';
import { FeatureToggle } from '../FeatureToggle';

describe('FeatureToggle Component', () => {
    it('should render children when feature is not on disabled list', () => {
        jest.spyOn(localStorage, 'getItem').mockImplementationOnce(() => "[]");

        const testContent = "testcontent";
        const view = render(<FeatureToggle feature={featuresEnum.FEATURE_1}>{testContent}</FeatureToggle>);

        expect(view.container).toMatchSnapshot();
        expect(view.container).toContainHTML(testContent);
    });

    it('should render empty template when feature is disabled and no offTemplate is present', () => {
      jest.spyOn(localStorage, 'getItem').mockImplementationOnce(() => "[\"FEATURE_1\"]");

      const testContent = "testcontent";
      const view = render(<FeatureToggle feature={featuresEnum.FEATURE_1}>{testContent}</FeatureToggle>);

      expect(view.container).toMatchSnapshot();
      expect(view.container).not.toContainHTML(testContent);
      expect(view.container).toBeEmptyDOMElement();
    });

    it('should render off template when feature is disabled and offTemplate is available', () => {
      jest.spyOn(localStorage, 'getItem').mockImplementationOnce(() => "[\"FEATURE_1\"]");

      const testContent = "testcontent";
      const testOffContent = "testoffcontent";
      const view = render(<FeatureToggle feature={featuresEnum.FEATURE_1} offTemplate={(<>{testOffContent}</>)}>{testContent}</FeatureToggle>);
      
      expect(view.container).toMatchSnapshot();
      expect(view.container).toContainHTML(testOffContent);
      expect(view.container).not.toContainHTML(testContent);
    });
});
