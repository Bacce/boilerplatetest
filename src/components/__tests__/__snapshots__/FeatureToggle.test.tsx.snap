// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureToggle Component should render children when feature is not on disabled list 1`] = `
<div>
  testcontent
</div>
`;

exports[`FeatureToggle Component should render empty template when feature is disabled and no offTemplate is present 1`] = `<div />`;

exports[`FeatureToggle Component should render off template when feature is disabled and offTemplate is available 1`] = `
<div>
  testoffcontent
</div>
`;
